File: leetcode/medium/133-cloneDeepGraph.js
// Given a reference of a node in a connected undirected graph.

// Return a deep copy (clone) of the graph.

// Each node in the graph contains a val (int) and a list (List[Node]) of its neighbors.

// class Node {
//     public int val;
//     public List<Node> neighbors;
// }
 
// Test case format:

// For simplicity sake, each node's value is the same as the node's index (1-indexed). For example, the first node with val = 1, 
// the second node with val = 2, and so on. The graph is represented in the test case using an adjacency list.

// Adjacency list is a collection of unordered lists used to represent a finite graph. Each list describes the set of neighbors of a 
// node in the graph.

// The given node will always be the first node with val = 1. You must return the copy of the given node as a reference to the cloned graph.

var cloneGraph = function(node) {
    if (!node) return null;
    
    let stack = [node];
    let visited = new Set();
    let graph = {};
    
    while (stack.length > 0) {
        let vertex = stack.pop();
        
        if (!(vertex.val in graph)) {
            graph[vertex.val] = new Node(vertex.val);
        }
        
        if (!visited.has(vertex.val)) {
            visited.add(vertex.val);
            
            for (let neighbor of vertex.neighbors) {
                if (!(neighbor.val in graph)) {
                    graph[neighbor.val] = new Node(neighbor.val);
                }
                graph[vertex.val].neighbors.push(graph[neighbor.val]);
                stack.push(neighbor);
            }
        }
    }
    return graph[node.val];
};

File: leetcode/medium/133-cloneDeepGraph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    public Node cloneGraph(Node node) {
        if (node == null) {
            return null;
        }
        
        Stack<Node> stack = new Stack<>();
        HashMap<Integer, Node> graph = new HashMap<>();
        Set<Integer> set = new HashSet<>();
        
        stack.add(node);
        
        while (!stack.isEmpty()) {
            Node currNode = stack.pop();
            
            if (graph.get(currNode.val) == null) {
                graph.put(currNode.val, new Node(currNode.val));
            }
            
            if (!set.contains(currNode.val)) {
                set.add(currNode.val);
                
                for (Node neighbor : currNode.neighbors) {
                    if (graph.get(neighbor.val) == null) {
                        graph.put(neighbor.val, new Node(neighbor.val));
                    }
                    graph.get(currNode.val).neighbors.add(graph.get(neighbor.val));
                    stack.push(neighbor);
                }
            }
        }
        return graph.get(node.val);
    }
}
